services/httpd/response_writer.go

```diff
@@ -147,7 +148,29 @@ func (w *csvFormatter) WriteResponse(resp Response) (n int, err error) {
 			}
 		}
 
-		for _, row := range result.Series {
+		for i, row := range result.Series {
+			if i > 0 && !reflect.DeepEqual(result.Series[i-1].Columns, row.Columns) {
+				// The columns have changed. Print a newline and reprint the header.
+				csv.Flush()
+				if err := csv.Error(); err != nil {
+					return n, err
+				}
+
+				out, err := io.WriteString(w, "\n")
+				if err != nil {
+					return n, err
+				}
+				n += out
+
+				w.columns = make([]string, 2+len(row.Columns))
+				w.columns[0] = "name"
+				w.columns[1] = "tags"
+				copy(w.columns[2:], row.Columns)
+				if err := csv.Write(w.columns); err != nil {
+					return n, err
```

Comment:
The `n` tracking is missing most of the csv calls - it's unfortunate that the csv.Writer doesn't expose the bytes written.

Assuming we do want to track `n` (for the QueryRequestBytesTransmitted stat), what do you think about using io.MultiWriter with the underlying `w` and some fake writer like

```go
type countWriter struct { n int64 }
func (w *countWriter) Write(p []byte) (int, err) {
  w.n += len(p)
  return len(p), nil
}
```

so that we can properly track `n`?

Reply:
That would be good. Can we open up another PR for that though? The behavior is already broken and I'm not breaking it anymore than it previously was.

Reply:
Sure thing - opened https://github.com/influxdata/influxdb/issues/9690 to track it.